examples/vision/image_captioning.py

```diff
@@ -0,0 +1,509 @@
+    output_sequence_length=SEQ_LENGTH,
+    standardize=custom_standardization,
+)
+vectorization.adapt(text_data)
+
+"""
+## Building a tf dataset pipeline for training
+
+We will generate pairs of images and the corresponding captions using `tf.data.Dataset`
+The pipeline consists of two steps:
+
+1. Read the image from the disk
+2. Tokenize all the five captions corresponding to this image
+"""
+
+
+def read_image(img_path, size=IMAGE_SIZE):
+    img = tf.io.read_file(img_path)
+    img = tf.image.decode_jpeg(img, channels=3)
+    img = tf.image.resize(img, IMAGE_SIZE)
+    return img
+
+
+def text_to_seq(cap):
+    return vectorization(cap)
+
+
+def make_dataset(images, captions):
+    img_dataset = tf.data.Dataset.from_tensor_slices(images).map(
+        read_image, num_parallel_calls=AUTOTUNE
+    )
+    cap_dataset = tf.data.Dataset.from_tensor_slices(captions).map(
+        text_to_seq, num_parallel_calls=AUTOTUNE
+    )
+    dataset = tf.data.Dataset.zip((img_dataset, cap_dataset))
+    dataset = dataset.batch(BATCH_SIZE).shuffle(256).prefetch(AUTOTUNE)
+    return dataset
+
+
+# Pass the list of images and the list of corresponding captions
+train_dataset = make_dataset(list(train_data.keys()), list(train_data.values()))
+valid_dataset = make_dataset(list(valid_data.keys()), list(valid_data.values()))
+
+"""
+## Building the model
+
+The image captioning task consists of three models in total. These are:
+
+1. CNN: Any CNN model as a backbone to extract the image features
+2. TransformerEncoder: The extracted image features are then passed to a Transformer
+                        based encodder that generates a new representation of the inputs
+3. TransformerDecoder: This model takes the encoder output and the text data (sequences)
+                        as inputs and tries to learn to generate the caption.
+"""
+
+
+def get_cnn_model():
+    base_model = efficientnet.EfficientNetB0(
+        input_shape=(*IMAGE_SIZE, 3), include_top=False, weights="imagenet",
+    )
+    # We freeze our feature extractor
+    base_model.trainable = False
+    base_model_out = base_model.output
+    base_model_out = layers.Reshape((-1, 1280))(base_model_out)
+    cnn_model = keras.models.Model(base_model.input, base_model_out)
+    return cnn_model
+
+
+class TransformerEncoderBlock(layers.Layer):
+    def __init__(self, embed_dim, dense_dim, num_heads, **kwargs):
+        super().__init__(**kwargs)
+        self.embed_dim = embed_dim
+        self.dense_dim = dense_dim
+        self.num_heads = num_heads
+        self.attention = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = layers.Dense(embed_dim, activation="relu")
+        self.layernorm_1 = layers.LayerNormalization()
+
+    def call(self, inputs, training, mask=None):
+        inputs = self.dense_proj(inputs)
+        attention_output = self.attention(
+            query=inputs, value=inputs, key=inputs, attention_mask=None
```

Comment:
We should also pass the padding mask here

Reply:
Is it really required here though? The model should be look at all the image features at once when applying attention to gather the global context for caption, right?

Reply:
That's a good point, there's no padding going on with the input sequence here so no masking necessary.

Reply:
Thanks for confirming